core/src/main/java/io/questdb/network/Socket.java

```diff
@@ -0,0 +1,114 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.network;
+
+import io.questdb.std.QuietCloseable;
+
+/**
+ * Abstraction for plain and encrypted TCP sockets. Encrypted sockets use additional buffer
+ * to accumulate messages, so they require extra calls to convert encrypted data to raw data.
+ */
+public interface Socket extends QuietCloseable {
+
+    /**
+     * @return file descriptor associated with the socket.
+     */
+    int getFd();
+
+    /**
+     * @return true if TLS session was already started.
+     */
+    boolean isTlsSessionStarted();
+
+    /**
+     * Reads encrypted data from the socket into internal buffer.
+     * No-op on plain socket.
+     *
+     * @return non-negative value if there were no errors.
+     */
+    int read();
+
+    /**
+     * Receives plain data into the given buffer from the socket. On encrypted
+     * sockets this call includes {@link #read()}, so an extra read() call
+     * is not required.
+     *
+     * @param bufferPtr pointer to the buffer
+     * @param bufferLen buffer length
+     * @return recv() result; non-negative if there were no errors.
+     */
+    int recv(long bufferPtr, int bufferLen);
+
+    /**
+     * Sends plain data from the given buffer to the socket. On encrypted
+     * sockets this call includes {@link #write()}, so an extra write() call
+     * is not required.
+     *
+     * @param bufferPtr pointer to the buffer
+     * @param bufferLen buffer length
+     * @return send() result; non-negative if there were no errors.
+     */
+    int send(long bufferPtr, int bufferLen);
+
+    /**
+     * Does a shutdown() call on the socket.
+     *
+     * @param how valid shutdown flag, e.g. {@link Net#SHUT_WR}.
+     * @return 0 if the call is successful; -1 if there was an error.
+     */
+    int shutdown(int how);
+
+    /**
+     * Starts a TLS session, if supported.
+     *
+     * @return 0 if the call is successful; -1 if there was an error.
+     */
+    int startTlsSession();
+
+    /**
+     * @return true if the socket support TLS encryption; false otherwise.
+     */
+    boolean supportsTls();
+
+    /**
+     * @return true if a {@link #read()} call should be made once
+     * the socket becomes readable.
+     */
+    boolean wantsRead();
+
+    /**
+     * @return true if a {@link #write()} call should be made once
+     * the socket becomes writable.
+     */
+    boolean wantsWrite();
+
+    /**
+     * Writes encrypted data from the internal buffer to the socket.
+     * No-op on plain socket.
+     *
+     * @return non-negative value if there were no errors.
+     */
+    int write();
```

Comment:
i find having `send()` and`write()` methods on the socket confusing. Can we keep only one?

Reply:
The `Socket` interface was simplified in e0d5a17261764d8e2671b183d2cfa6511b7680fe

Both TLS methods are now merged into a single one: `tlsIO`. `recv` and `send` remain.